File: src/main/java/me/kikugie/techutils/render/litematica/LitematicSpace.java
package me.kikugie.techutils.render.litematica;

import fi.dy.masa.litematica.schematic.LitematicaSchematic;
import net.minecraft.util.math.Box;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

public class LitematicSpace {
    public final Map<String, LitematicRegionBlockView> regionBlockViewMap;
    public final Map<String, Box> boxes;
    public final Box fullBox;

    public LitematicSpace(LitematicaSchematic schematic) {
        this.boxes = getConvertedBoxes(schematic.getAreas());
        this.fullBox = getFullBox(boxes);
        this.regionBlockViewMap = getRegionBlockViewMap(schematic);
    }

    private Map<String, LitematicRegionBlockView> getRegionBlockViewMap(LitematicaSchematic schematic) {
        Collection<String> areas = schematic.getAreas().keySet();
        HashMap<String, LitematicRegionBlockView> containers = new HashMap<>(areas.size());
        areas.forEach(name -> containers.put(name, new LitematicRegionBlockView(schematic.getSubRegionContainer(name), boxes.get(name))));
        return containers;
    }

    private Map<String, Box> getConvertedBoxes(Map<String, fi.dy.masa.litematica.selection.Box> areas) {
        HashMap<String, Box> boxes = new HashMap<>(areas.size());
        areas.forEach((name, region) -> boxes.put(name, new Box(region.getPos1(), region.getPos2())));
        return boxes;
    }

    private Box getFullBox(Map<String, Box> boxes) {
        int[] dims = {0, 0, 0, 0, 0, 0};
        boxes.forEach((name, box) -> {
            dims[0] = Math.min(dims[0], (int) box.minX);
            dims[1] = Math.min(dims[1], (int) box.minY);
            dims[2] = Math.min(dims[2], (int) box.minZ);
            dims[3] = Math.max(dims[3], (int) box.maxX);
            dims[4] = Math.max(dims[4], (int) box.maxY);
            dims[5] = Math.max(dims[5], (int) box.maxZ);
        });
        return new Box(dims[0], dims[1], dims[2], dims[3], dims[4], dims[5]);
    }
}

File: src/main/kotlin/me/kikugie/techutils/render/litematica/LitematicSpace.kt
package me.kikugie.techutils.render.litematica

import fi.dy.masa.litematica.schematic.LitematicaSchematic
import fi.dy.masa.litematica.selection.Box
import net.minecraft.util.math.*
import java.util.function.Consumer
import kotlin.math.max
import kotlin.math.min

class LitematicSpace(schematic: LitematicaSchematic?) {
    val regionBlockViewMap: Map<String?, LitematicRegionBlockView?>
    val boxes: Map<String?, net.minecraft.util.math.Box?>
    val fullBox: net.minecraft.util.math.Box

    init {
        boxes = getConvertedBoxes(schematic!!.areas)
        fullBox = getFullBox(boxes)
        regionBlockViewMap = getRegionBlockViewMap(schematic)
    }

    private fun getRegionBlockViewMap(schematic: LitematicaSchematic?): Map<String?, LitematicRegionBlockView?> {
        val areas: Collection<String> = schematic!!.areas.keys
        val containers = HashMap<String?, LitematicRegionBlockView?>(areas.size)
        areas.forEach(Consumer { name: String? ->
            containers[name] = LitematicRegionBlockView(
                schematic.getSubRegionContainer(name), boxes[name]
            )
        })
        return containers
    }

    private fun getConvertedBoxes(areas: Map<String, Box>): Map<String?, net.minecraft.util.math.Box?> {
        val boxes = HashMap<String?, net.minecraft.util.math.Box?>(areas.size)
        areas.forEach { (name: String?, region: Box) -> boxes[name] = Box(region.pos1, region.pos2) }
        return boxes
    }

    private fun getFullBox(boxes: Map<String?, net.minecraft.util.math.Box?>): net.minecraft.util.math.Box {
        val dims = intArrayOf(0, 0, 0, 0, 0, 0)
        boxes.forEach { (_: String?, box: net.minecraft.util.math.Box?) ->
            dims[0] = min(dims[0], box!!.minX.toInt())
            dims[1] = min(dims[1], box.minY.toInt())
            dims[2] = min(dims[2], box.minZ.toInt())
            dims[3] = max(dims[3], box.maxX.toInt())
            dims[4] = max(dims[4], box.maxY.toInt())
            dims[5] = max(dims[5], box.maxZ.toInt())
        }
        return Box(
            dims[0].toDouble(),
            dims[1].toDouble(),
            dims[2].toDouble(),
            dims[3].toDouble(),
            dims[4].toDouble(),
            dims[5].toDouble()
        )
    }
}